optuna/search_space/intersection.py

```diff
@@ -17,11 +17,6 @@ class IntersectionSearchSpace:
     neither is included in the resulting search space
     (i.e., the parameters with dynamic value ranges are excluded).
 
-    Note that an instance of this class is supposed to be used for only one study.
```

Comment:
I think this is still true since `self._cursor` and `self._search_space` are still associated with a particular study. I suggest we keep this line, and if to keep this PR no compat breaking, reintroduce the `raise ValueError("IntersectionSearchSpace cannot handle multiple studies.")` at the callers, e.g. in the samplers instances, store the study ID, and raise this error. This behavior itself is not so pretty in my opinion, and can be discussed, though. What do you think?

Reply:
I agree. Regarding the current design, developers of user-defined samplers need to manage relationship between `IntersectionSearchSpace` instances and studies. Otherwise, `IntersectionSearchSpace.calculate` will return unexpected values silintly.

How about sticking to the `study` as an argument of `IntersectionSearchSpace.calculate`?
I roughly implemented a prototype in the commit https://github.com/optuna/optuna/commit/c8b8e0db5917d3860eee39a9555729d75d0b3f51.

Reply:
Thank you for your comments and insightful prototype! I fixed my PR to keep the study as an argument of `IntersectionSearchSpace.calculate` and change the argument of `intersection_search_space`.

 As a difference to prototype by @toshihikoyanase , I does not extract common interface from `IntersectionSearchSpace.calculate` and `intersection_search_space`. Instead, I implement some required part of of `IntersectionSearchSpace.calculate` to `intersection_search_space`. This violates DRY principle, but I think this change makes some code simpler. What do you think about this change?

Reply:
Thank you for your update. I'm a bit afraid of the maintenance of the two similar logics. Developers need to synchronize the change in both search space calculation logics, and the review cost may be doubled since they are similar, but not exactly the same.

What do you think of extracting a part of the calculation logic into a function?
The return type of `_calculate` in https://github.com/optuna/optuna/commit/c8b8e0db5917d3860eee39a9555729d75d0b3f51 was complicated since it created both external and internal search space. If we return the internal search space, code may be more readable.

https://github.com/optuna/optuna/compare/master...toshihikoyanase:optuna:ya-4514?expand=1

Reply:
Thank you for your comment. I update my code according to direction of extracting a part of the calculation logic. I implemented myself, but eventually, it becomes almost same as your prototype.

I have 2 points to consult.
1. This update changes the interface of `intersection_search_space` function. Now, the interface is different from `IntersectionSearchSpace.calculate`. So, should we change the name of this function? For example, my idea is `intersection_search_space_from_trials`.
2. (nits) When annotating a variable that can take None or other types, which is better to use Optional or |?

Reply:
> So, should we change the name of this function?

Good point! I don't think it is necessary to rename the function. This is because 
- we provide type hints, and modern development environments can notify developers
- I expect `intersection_search_space` if `intersection_search_space_from_trials` exists, but it may a bit overengineering

> (nits) When annotating a variable that can take None or other types, which is better to use Optional or |?

I we apply `from __future__ import annotations` to code, `|` seems to be selected as can be seen in #4526. But, I think we can work on it in a follow-up PR.

Reply:
Ok, I leave the function name as now.
And I updated my PR to use `|` instead of `Optional`.

Reply:
Thanks!